compiler-core/src/build/module_loader.rs

```diff
@@ -49,26 +80,34 @@ where
         let artefact = name.replace("/", "@");
         let source_mtime = self.io.modification_time(&path)?;
 
-        let read_source = |name| Ok(Input::New(self.read_source(path, name, source_mtime)?));
-
         let meta = match self.read_cache_metadata(&artefact)? {
             Some(meta) => meta,
-            None => return read_source(name),
+            None => return self.read_source(path, name, source_mtime).map(Input::New),
         };
 
-        // If there's a timestamp and it's newer or the same than the source
-        // file modification time then we read the cached data.
+        let mut module = None;
+
+        // If the timestamp of the source is newer than the cache entry and
+        // the hash of the source differs from the one in the cache entry,
+        // then we need to recompile.
         if meta.mtime < source_mtime {
-            tracing::debug!(?name, "cache_stale");
-            return read_source(name);
+            let source_module = self.read_source(path.clone(), name.clone(), source_mtime)?;
+            if meta.digest != SourceDigest::new(&source_module.code) {
+                tracing::debug!(?name, "cache_stale");
+                return Ok(Input::New(source_module));
+            }
+            module = Some(source_module);
         }
 
         // The cache currently does not contain enough data to perform codegen,
         // so if codegen is required in this compiler run then we must check
         // that codegen has already been performed before using a cache.
         if self.codegen.is_required() && !meta.codegen_performed {
             tracing::debug!(?name, "codegen_required_cache_insufficient");
-            return read_source(name);
+            return match module {
+                Some(m) => Ok(Input::New(m)),
+                None => self.read_source(path, name, source_mtime).map(Input::New),
+            };
```

Comment:
This method has become a lot harder to read, going from a sequence to checks to a function with nesting and the src value shared between the branches.

If it can be flat again that would be my preference, if not let's commit to nested conditional logic and move this into the branch above, so that the logic is written in a consistent style and we can remove the `module` variable.

Reply:
I was able to get rid of this variable by reordering if statements. Now we check for codegen case first and then we check for modification time. The code is flat now and I think that semantics is fine, but let me know if this is not the case.